pennylane_sf/gbs.py

```diff
@@ -0,0 +1,258 @@
+                device
+            observables (list[~pennylane.operation.Operation]): observables to be measured
+            variable_deps (dict[int, ParameterDependency]): reference dictionary
+                mapping free parameter values to the operations that
+                depend on them
+
+        Returns:
+            array[float]: Jacobian matrix of size (``len(probs)``, ``num_wires``)
+        """
+        self.reset()
+
+        requested_wires = observables[0].wires
+
+        # Create dummy observable to measure probabilities over all wires
+        obs_all_wires = qml.Identity(wires=self.wires)
+        obs_all_wires.return_type = Probability
+        prob = self.execute(operations, [obs_all_wires], parameters=variable_deps)[0]
+
+        jac = self._jacobian_all_wires(prob)
+
+        if requested_wires == self.wires:
+            return jac
+
+        # Unflatten into a [cutoff, cutoff, ..., cutoff, num_params] dimensional tensor
+        jac = jac.reshape([self.cutoff] * self.num_wires + [self.num_wires])
+
+        # Find indices to trace over
+        all_indices = set(range(self.num_wires))
+        requested_indices = set(self.wires.indices(requested_wires))
+        trace_over_indices = all_indices - requested_indices
+
+        traced_jac = np.sum(jac, axis=tuple(trace_over_indices))
+
+        # Flatten into [cutoff ** num_requested_wires, num_params] dimensional tensor
+        traced_jac = traced_jac.reshape(-1, self.num_wires)
+
+        return traced_jac
+
+    def _jacobian_all_wires(self, prob):
+        """Calculates the jacobian of the probability distribution with respect to all wires.
+
+        This function uses Eq. (28) of `this <https://arxiv.org/pdf/2004.04770.pdf>`__ paper.
+
+        Args:
+            prob (array[float]): the probability distribution as a flat array
+
+        Returns:
+            array[float]: the jacobian
+        """
+        jac = np.zeros([len(prob), self.num_wires])
+
+        n = len(self._WAW)
+        disp = np.zeros(2 * n)
+        cov = self.calculate_covariance(self._WAW, hbar=self.hbar)
+        mean_photons_by_mode = photon_number_mean_vector(disp, cov, hbar=self.hbar)
+
+        for i, s in enumerate(np.ndindex(*[self.cutoff] * self.num_wires)):
+            jac[i] = (s - mean_photons_by_mode) * prob[i] / self._params
+
+        return jac
+
+    @staticmethod
+    def calculate_covariance(A, hbar):
+        r"""Calculates the covariance matrix corresponding to an input adjacency matrix.
+
+        Args:
+            A (array[float]): adjacency matrix
+            hbar (float): the convention chosen in the canonical commutation relation
+                :math:`[x, p] = i \hbar`
+
+        Returns:
+            array[float]: covariance matrix
+        """
+        n = len(A)
+        I = np.identity(2 * n)
+        o_mat = np.block([[np.zeros_like(A), np.conj(A)], [A, np.zeros_like(A)]])
+        cov = hbar * (np.linalg.inv(I - o_mat) - I / 2)
+        return cov
```

Comment:
Weird that we don't have a function in The Walrus for this (@nquesada?) Using The Walrus, it looks like you would have to do

```python
>>> Covmat(gen_Qmat_from_graph(A, n_mean), hbar)
```

Reply:
@thisac this would be nice to consider in The Walrus quantum refactor. Aside from organization, it would also be good to standardize naming for utility conversion functions :)